File: sungwoopark95/Algorithms/bellman_ford.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <limits>
using namespace std;

class Vertex {
public:
    double d = numeric_limits<double>::infinity(); // v와 s 사이의 거리를 나타냅니다.
    Vertex* pi = nullptr; // v의 부모 노드를 나타냅니다.
};

class Graph {
public:
    vector<Vertex*> V; // 정점의 집합
    vector<pair<Vertex*, Vertex*>> E; // 간선의 집합
};

struct pair_hash {
    template <class T1, class T2>
    size_t operator () (const pair<T1,T2>& p) const {
        auto h1 = hash<T1>{}(p.first);
        auto h2 = hash<T2>{}(p.second);
        return h1 ^ h2;
    }
};

void INITIALIZE_SINGLE_SOURCE(Graph& G, Vertex* s) {
    for (Vertex* v : G.V) {
        v->d = numeric_limits<double>::infinity();
        v->pi = nullptr;
    }
    s->d = 0;
}

void RELAX(Vertex* u, Vertex* v, unordered_map<pair<Vertex*, Vertex*>, double, pair_hash>& weights) {
    if (weights.find({u, v}) != weights.end() && v->d > u->d + weights[{u, v}]) {
        v->d = u->d + weights[{u, v}];
        v->pi = u;
    }
}

vector<Vertex*> construct_path(Vertex* s, Vertex* v) {
    vector<Vertex*> path;
    while (v != s && v != nullptr) {
        path.insert(path.begin(), v);
        v = v->pi;
    }
    if (v == s) {
        path.insert(path.begin(), s);
    }
    return path;
}

unordered_map<Vertex*, vector<Vertex*>> BELLMAN_FORD(Graph& G, unordered_map<pair<Vertex*, Vertex*>, double, pair_hash>& weights, Vertex* s) {
    INITIALIZE_SINGLE_SOURCE(G, s);
    
    for (int i = 0; i < G.V.size() - 1; ++i) {
        for (auto edge : G.E) {
            RELAX(edge.first, edge.second, weights);
        }
    }
    
    for (auto edge : G.E) {
        if (edge.second->d > edge.first->d + weights[edge]) {
            return {}; // 음의 순환이 있으면 빈 맵을 반환합니다.
        }
    }
    
    unordered_map<Vertex*, vector<Vertex*>> shortest_paths;
    for (Vertex* v : G.V) {
        shortest_paths[v] = construct_path(s, v);
    }
    
    return shortest_paths;
}

int main() {
    // 그래프 초기화
    Vertex v1, v2, v3;
    Graph G;
    G.V.push_back(&v1);
    G.V.push_back(&v2);
    G.V.push_back(&v3);
    G.E.push_back({&v1, &v2});
    G.E.push_back({&v2, &v3});
    
    unordered_map<pair<Vertex*, Vertex*>, double, pair_hash> weights;
    weights[make_pair(&v1, &v2)] = 5;
    weights[make_pair(&v2, &v3)] = 3;

    // 알고리즘 실행
    auto paths = BELLMAN_FORD(G, weights, &v1);

    // 결과 출력
    for (auto& [vertex, path] : paths) {
        cout << "Path to vertex " << vertex << ": ";
        for (Vertex* v : path) {
            cout << v << " ";
        }
        cout << endl;
    }

    return 0;
}

// negative cycle이 있는지를 확인하는 코드
// bool BELLMAN_FORD(Graph& G, unordered_map<pair<Vertex*, Vertex*>, double, pair_hash>& weights, Vertex* s) {
//     INITIALIZE_SINGLE_SOURCE(G, s);
    
//     for (int i = 0; i < G.V.size() - 1; ++i) {
//         for (auto edge : G.E) {
//             RELAX(edge.first, edge.second, weights);
//         }
//     }
    
//     for (auto edge : G.E) {
//         if (edge.second->d > edge.first->d + weights[edge]) {
//             return false; // 음의 순환이 있으면 false를 반환합니다.
//         }
//     }
    
//     return true; // 그렇지 않으면 true를 반환합니다.
// }

// int main() {
//     // 그래프 초기화
//     Vertex v1, v2, v3;
//     Graph G;
//     G.V.push_back(&v1);
//     G.V.push_back(&v2);
//     G.V.push_back(&v3);
//     G.E.push_back({&v1, &v2});
//     G.E.push_back({&v2, &v3});
    
//     unordered_map<pair<Vertex*, Vertex*>, double, pair_hash> weights;
//     weights[make_pair(&v1, &v2)] = 5;
//     weights[make_pair(&v2, &v3)] = 3;

//     // 알고리즘 실행
//     bool result = BELLMAN_FORD(G, weights, &v1);

//     // 결과 출력
//     if (result) {
//         cout << "No negative cycles detected." << endl;
//     } else {
//         cout << "Negative cycle detected!" << endl;
//     }

//     return 0;
// }

File: sungwoopark95/python/Algorithms/bellmanford.py
from __future__ import annotations
from typing import List, Dict, Tuple

class WeightedEdge:
    def __init__(self, end:Vertex, weight:int) -> None:
        self.end = end
        self.weight = weight


class Vertex:
    def __init__(self, name:str) -> None:
        self.name = name
        self.key = 0
        self.parent = None
        self.adjacency_list = []
    
    def GetAdjacencyList(self) -> List[WeightedEdge]:
        return self.adjacency_list
    
    def AddToAdjacentList(self, vertex:Vertex, weight:int) -> None:
        edge = WeightedEdge(end=vertex, weight=weight)
        self.adjacency_list.append(edge)
        
        
class Graph:
    def __init__(self) -> None:
        self.vertices = []
    
    def GenVertex(self, name) -> Vertex:
        v = Vertex(name)
        self.vertices.append(v)
        return v
    
    def GenEdge(self, start:Vertex, end:Vertex, weight:int) -> None:
        start.AddToAdjacentList(vertex=end, weight=weight)
        
    def GetEdge(self) -> List[WeightedEdge]:
        edges = []
        for v in self.vertices:
            for adj in v.GetAdjacencyList():
                edges.append((v, adj.end, adj.weight))
        return edges
    
    def GetVertices(self) -> List[Vertex]:
        return self.vertices
    
    def GetNumVertices(self) -> int:
        return len(self.vertices)
    
    def PrintGraph(self) -> None:
        print("============== Defined Graph ==============")
        for vertex in self.vertices:
            print(f"Vertex {vertex.name}, Adjacency List: ", end='')
            is_first = True
            for adj in vertex.GetAdjacencyList():
                # adj : WeightedEdge object
                if not is_first:
                    print(", ", end='')
                print(f"(Vertex {adj.end.name}, Weight {adj.weight})", end='')
                is_first = False
            print()
    
    def __del__(self) -> None:
        ## unnecessary for python
        for v in self.vertices:
            del v
        del self.vertices


def initialize_single_source(G:Graph, start:Vertex) -> None:
    for vertex in G.GetVertices():
        vertex.key = float('inf')
        vertex.parent = None
    start.key = 0


def relax(u:Vertex, v:Vertex, w:int) -> None:
    if v.key > u.key + w:
        v.key = u.key + w
        v.parent = u


def construct_path(u:Vertex, v:Vertex) -> List[Vertex]:
    """u에서 v까지의 경로를 반환합니다."""
    path = []
    while v != u and v is not None:
        path.insert(0, v)
        v = v.parent
    if v == u:
        path.insert(0, u)
    return path


def bellman_ford_check_cycle(G:Graph, start:Vertex) -> bool:
    initialize_single_source(G, start)
    
    for _ in range(1, G.GetNumVertices()):
        for edge in G.GetEdge():
            u, v, w = edge
            relax(u, v, w)
    
    # check negative cycle
    for edge in G.GetEdge():
        u, v, w = edge
        if v.key > u.key + w:
            return False
    return True


def bellman_ford_mst(G:Graph, start:Vertex) -> List[Tuple[Vertex]]:
    initialize_single_source(G, start)

    for _ in range(1, G.GetNumVertices()):
        for edge in G.GetEdge():
            u, v, w = edge
            relax(u, v, w)
    
    mst = []
    for v in G.GetVertices():
        if v.parent:
            edge = (v.parent.name, v.name)
            mst.append(edge)
    return mst


def bellman_ford_shortest_paths(G:Graph, start:Vertex) -> Dict[Vertex, List[Vertex]]:
    shortest_paths = dict()
    initialize_single_source(G, start)
    for _ in range(1, G.GetNumVertices()):
        for edge in G.GetEdge():
            u, v, w = edge
            relax(u, v, w)
    
    for v in G.GetVertices():
        shortest_paths[v] = construct_path(start, v)        
    return shortest_paths


if __name__ == "__main__":
    graph = Graph()
    
    # generate vertices
    vertex_1 = graph.GenVertex('1')
    vertex_2 = graph.GenVertex('2')
    vertex_3 = graph.GenVertex('3')
    vertex_4 = graph.GenVertex('4')
    vertex_5 = graph.GenVertex('5')
    
    # generate edges
    graph.GenEdge(vertex_1, vertex_2, 6)
    graph.GenEdge(vertex_1, vertex_3, 1)
    graph.GenEdge(vertex_1, vertex_4, 4)
    graph.GenEdge(vertex_1, vertex_5, 6)
    graph.GenEdge(vertex_3, vertex_5, 1)
    graph.GenEdge(vertex_4, vertex_2, 2)
    graph.GenEdge(vertex_4, vertex_5, 4)
    graph.GenEdge(vertex_5, vertex_2, 3)

    # print all vertices and their adjacency lists
    graph.PrintGraph()
    
    # check if there exists any negative cycle in the graph
    print(f"Is this graph acyclic? : {bellman_ford_check_cycle(graph, vertex_1)}")
    
    start = vertex_1
    print(f"Minimum spanning tree of this graph : {bellman_ford_mst(graph, start)}")
    shortest_paths = bellman_ford_shortest_paths(graph, start)
    for key in shortest_paths:
        print(f"Shortest path from {start.name} to {key.name}: ", end='')
        is_start = True
        for item in shortest_paths[key]:
            if not is_start:
                print(" -> ", end='')
            print(item.name, end='')
            is_start = False
        print()
            